src/nvenc/nvenc_base.cpp

```diff
@@ -0,0 +1,541 @@
+    pic_params.encodePicFlags = force_idr ? NV_ENC_PIC_FLAG_FORCEIDR : 0;
+    pic_params.inputTimeStamp = frame_index;
+    pic_params.pictureStruct = NV_ENC_PIC_STRUCT_FRAME;
+    pic_params.inputBuffer = mapped_input_buffer.mappedResource;
+    pic_params.bufferFmt = mapped_input_buffer.mappedBufferFmt;
+    pic_params.outputBitstream = output_bitstream;
+    pic_params.completionEvent = async_event_handle;
+
+    if (nvenc_failed(nvenc->nvEncEncodePicture(encoder, &pic_params))) {
+      BOOST_LOG(error) << "NvEncEncodePicture failed: " << last_error_string;
+      return {};
+    }
+
+    NV_ENC_LOCK_BITSTREAM lock_bitstream = { NV_ENC_LOCK_BITSTREAM_VER };
+    lock_bitstream.outputBitstream = output_bitstream;
+    lock_bitstream.doNotWait = 0;
+
+    if (async_event_handle && !wait_for_async_event(100)) {
+      BOOST_LOG(error) << "NvEnc: frame " << frame_index << " encode wait timeout";
+      return {};
+    }
+
+    if (nvenc_failed(nvenc->nvEncLockBitstream(encoder, &lock_bitstream))) {
+      BOOST_LOG(error) << "NvEncLockBitstream failed: " << last_error_string;
+      return {};
+    }
+
+    auto data_pointer = (uint8_t *) lock_bitstream.bitstreamBufferPtr;
+    nvenc_encoded_frame encoded_frame {
+      { data_pointer, data_pointer + lock_bitstream.bitstreamSizeInBytes },
+      lock_bitstream.outputTimeStamp,
+      lock_bitstream.pictureType == NV_ENC_PIC_TYPE_IDR,
+      encoder_state.rfi_needs_confirmation,
+    };
+
+    if (encoder_state.rfi_needs_confirmation) {
+      // Invalidation request has been fulfilled, and video network packet will be marked as such
+      encoder_state.rfi_needs_confirmation = false;
+    }
+
+    encoder_state.last_encoded_frame_index = frame_index;
+
+    if (encoded_frame.idr) {
+      BOOST_LOG(debug) << "NvEnc: idr frame " << encoded_frame.frame_index;
+    }
+
+    if (nvenc_failed(nvenc->nvEncUnlockBitstream(encoder, lock_bitstream.outputBitstream))) {
+      BOOST_LOG(error) << "NvEncUnlockBitstream failed: " << last_error_string;
+    }
+
+    if (config::sunshine.min_log_level <= 1) {
+      // Print encoded frame size stats to debug log every 20 seconds
+      auto callback = [&](float stat_min, float stat_max, double stat_avg) {
+        auto f = stat_trackers::one_digit_after_decimal();
+        BOOST_LOG(debug) << "NvEnc: encoded frame sizes (min max avg) " << f % stat_min << " " << f % stat_max << " " << f % stat_avg << " kB";
+      };
+      using namespace std::literals;
+      encoder_state.frame_size_tracker.collect_and_callback_on_interval(encoded_frame.data.size() / 1000., callback, 20s);
+    }
+
+    return encoded_frame;
+  }
+
+  bool
+  nvenc_base::invalidate_ref_frames(uint64_t first_frame, uint64_t last_frame) {
+    if (!encoder || !encoder_params.rfi) return false;
+
+    if (last_frame < first_frame ||
+        encoder_state.last_encoded_frame_index < first_frame ||
+        encoder_state.last_encoded_frame_index > first_frame + 100) {

```

Comment:
That last condition (`encoder_state.last_encoded_frame_index > first_frame + 100`) is actually possible if we have a period of severe packet loss. We will continue expanding our RFI range until we successfully receive a recovery frame from the host. We shouldn't log an error for that case.

Reply:
The reason why it's there is because nvenc rfi api doesn't work with ranges, but individual frames. And these calls can get expensive really fast if we don't limit them to some reasonable number. I will add this to my todo, some more intelligent logic can be written based on previously requested range, but it shouldn't be a blocker since it's just a log message that happens in very specific circumstances.

Reply:
That's fine with me. I just mentioned it because I saw the logs when I was performing RFI and packet loss tests with this branch.